api/v3/apierrors/errors.go

```diff
@@ -0,0 +1,197 @@
+	var source string
+	if err := json.Unmarshal(data, &source); err != nil {
+		return err
+	}
+	*i = ToInvalid(source)
+	return nil
+}
+
+// InvalidParameter is a single field that failed input validation.
+type InvalidParameter struct {
+	// Field concerned by the error.
+	Field string `json:"field"`
+
+	// Rule represents the rule that has triggered the error.
+	Rule string `json:"rule,omitempty"`
+
+	// Reason describes why the error has been triggered.
+	Reason string `json:"reason"`
+
+	// Source describes where the error has been triggered: body, header, path.
+	Source InvalidParameterSource `json:"source"`
+
+	// Choices represents the available choices for value in a case of an enum.
+	Choices []string `json:"choices,omitempty"`
+
+	// Minimum is an optional field for setting the minimum required value for
+	// an attribute.
+	Minimum *int `json:"minimum,omitempty"`
+
+	// Maximum is an optional field for setting the maximum required value for
+	// an attribute.
+	Maximum *int `json:"maximum,omitempty"`
+
+	// Dependents is an optional field for when the rule "dependent_fields" is
+	// applied.
+	Dependents []string `json:"dependents,omitempty"`
+}
+
+// Stringer method for a collection of InvalidParameter entities.
+func (ips InvalidParameters) String() string {
+	out := new(strings.Builder)
+	for i, param := range ips {
+		out.WriteString(param.Field)
+		if param.Rule != "" {
+			_, _ = fmt.Fprintf(out, " [%s]", param.Rule)
+		}
+		_, _ = fmt.Fprintf(out, ": %s", param.Reason)
+		if i != len(ips)-1 {
+			_, _ = fmt.Fprintf(out, ", ")
+		}
+	}
+	return out.String()
+}
+
+// Error satisfies the error interface.
+func (bae *BaseAPIError) Error() string {
+	switch {
+	case bae.InvalidParameters != nil:
+		return fmt.Sprintf("%s: %s", bae.UnderlyingError, bae.InvalidParameters.String())
+	case bae.Detail != "" && bae.UnderlyingError != nil:
+		return fmt.Sprintf("%s: %s", bae.Detail, bae.UnderlyingError)
+	case bae.Detail != "":
+		return bae.Detail
+	}
+	return bae.UnderlyingError.Error()
+}
+
+// Unwrap returns the underlying error
+func (bae *BaseAPIError) Unwrap() error {
+	return bae.UnderlyingError
+}
+
+// Context is the context that created the error
+func (bae *BaseAPIError) Context() context.Context {
+	return bae.ctx
+}
+
+// HandleAPIError is a helper function that accepts an error
+func (bae *BaseAPIError) HandleAPIError(
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	_ = render.RenderJSON(w, bae, render.WithContentType(ContentTypeProblemValue), render.WithStatus(bae.Status))
+}
+
+// GenericErrorEncoder is an error encoder that encodes the error as a generic error.
+func GenericErrorEncoder() encoder.ErrorEncoder {
+	return func(ctx context.Context, err error, w http.ResponseWriter, r *http.Request) bool {
+		if err, ok := lo.ErrorsAs[*BaseAPIError](err); ok {
+			err.HandleAPIError(w, r)
+			return true
+		}
+
+		return commonhttp.HandleIssueIfHTTPStatusKnown(ctx, err, w)
+	}
+}
```

Comment:
Let's move this at least to a seperate file, to help with adopting the opensourced kong libs if needed.

I would also consolidate pkg/framework/commonhttp/encoder.go:GenericErrorEncoder and add those handlers here with proper kong compliant error formats for the generic errors at least.